api_loader.go

```diff
@@ -584,32 +582,23 @@ func loadApps(specs []*APISpec, muxer *mux.Router) {
 	}
 
 	for i, spec := range specs {
-		go func(spec *APISpec, i int) {
-			subrouter := hostRouters[spec.Domain]
-			if subrouter == nil {
-				mainLog.WithFields(logrus.Fields{
-					"domain": spec.Domain,
-					"api_id": spec.APIID,
-				}).Warning("Trying to load API with Domain when custom domains are disabled.")
-				subrouter = muxer
-			}
-
-			chainObj := processSpec(spec, apisByListen, redisStore, redisOrgStore, healthStore, rpcAuthStore, rpcOrgStore, subrouter, logrus.NewEntry(log))
+		subrouter := hostRouters[spec.Domain]
+		if subrouter == nil {
+			mainLog.WithFields(logrus.Fields{
+				"domain": spec.Domain,
+				"api_id": spec.APIID,
+			}).Warning("Trying to load API with Domain when custom domains are disabled.")
+			subrouter = muxer
+		}
 
-			chainObj.Index = i
-			chainChannel <- chainObj
-			apisMu.Lock()
-			spec.middlewareChain = chainObj
-			apisMu.Unlock()
-		}(spec, i)
```

Comment:
just something draw my attention - here we possibly had the same spec passed to go-routine because for-cycle vars are being shared between iterations

Reply:
Maybe, but in general, it is very unsafe thing if you run on the multi-core machine and it schedules these goroutines in parallel.